### DISCUSSION TITLE
Result analysis with Python and Pandas

### DISCUSSION THREAD

**Discussion by carlosmn25** [omnetpp/omnetpp#1189] [Q&A] [answered]
Hi all,
I'm currently working with OMNeT++ 6.0.1. As the [Result Analysis with Python](https://docs.omnetpp.org/tutorials/pandas/) tutorial says, I've noticed that the IDE analysis tools are useful for quick visualization, but sometimes limiting when it comes to automation, extensive analysis or fine-grained customization of graphs. Thus, I'd like to analyze the results with Pandas to create my own graphs. In the tutorial I cited above, it says that it's obsolete since it was written for OMNeT++ 5.X and OMNeT++ 6.X is supposed to provide tools for direct data parsing and manipulation from Python. Also, I've found [these slides](https://summit.omnetpp.org/2022/assets/pdf/pres-pyscr.pdf) from the OMNeT++ summit in 2022 where an "Analysis API" for Python is mentioned as WIP.

However, I couldn't find in the documentation how to use/install this API or if it is already available. This is possibly my fault, maybe I didn't search properly or enough, but I was wondering if someone could provide me with some documentation, guidelines or tips on what is the "state-of-the-art" way of analyzing results with Python outside the IDE. Should I just use the scave tool to convert files into CSV and then read them with Pandas, or is there a more direct way of doing it inside python by installing some package?
Thanks in advance :)

**Comment by rhornig** [comment, marked as answer]
In 6.0, the result analysis inside the IDE is also using Python. In fact all the fancy controls in the IDE just generate a python file at the end and the results are displayed in the IDE. You can modify the generated python files manually (or start your own).

There is an API in the omnetpp_root/python folder under the omnetpp.scave package, that uses the command line scave tool t export the data to csv files and then read them in python into data frames. At that point you can do anything you would like from the command line python script.

(if you run your python script from the omnet environment (i.e. after setenv), that dirytory is already added to the python path, so a simple `import omnetpp.scave` should be enough. As a reference, the test/scave/results_api folder contains all the tests for the scave result API.

In a few days omnetpp 6.1rc will be released where the scave command line tool is no longer used for result reading. reading is done using native OMNeT++ resultfile readers, which will give you performance benefit compared to the 6.0.3 version.

Also, the analysis chapter in the manual got updated to cover the python APi usage.

https://github.com/omnetpp/omnetpp/blob/master/doc/src/manual/ch-ana-sim.tex#L863

**Reply by carlosmn25** [reply]
Thanks a lot! I knew that the IDE controls use Python underneath, but I want to use Python myself. I think the API references that you gave me will be of great help.

**Reply by rhornig** [reply]
Just on time. omnetpp-6.1rc1 just released. The omnetpp.scave package in omnetpp_root/python is  now loading result data directly using native c binding. If you run your python script inside the omnet++ setenv environment the module should be available for you for importing.
